libmamba/src/core/download_impl.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_CORE_DOWNLOAD_IMPL_HPP
+#define MAMBA_CORE_DOWNLOAD_IMPL_HPP
+
+#include <chrono>
+#include <unordered_map>
+
+#include "mamba/core/download.hpp"
+
+#include "compression.hpp"
+#include "curl.hpp"
+
+namespace mamba
+{
+    enum class DownloadState
+    {
+        WAITING,
+        PREPARING,
+        RUNNING,
+        FINISHED,
+        FAILED
+    };
+
+    class DownloadAttempt
+    {
+
+    public:
+
+        using completion_function = std::function<bool(CURLMultiHandle&,CURLcode)>;
+        using on_success_callback = std::function<bool(DownloadSuccess)>;
+        using on_failure_callback = std::function<bool(DownloadError)>;
+
+        DownloadAttempt(CURLHandle& handle, const DownloadRequest* request);
```

Comment:
As pointed before, request should be taken by const reference, passing nullptr should not be authorized.